Tools/build/check_warnings.py

```diff
@@ -149,44 +183,53 @@ def main(argv: list[str] | None = None) -> int:
     # Check that the compiler output file is a valid path
     if not Path(args.compiler_output_file_path).is_file():
         print(
-            "Compiler output file does not exist: "
-            f"{args.compiler_output_file_path}"
+            f"Compiler output file does not exist: {args.compiler_output_file_path}"
         )
         return 1
 
-    # Check that the warning ignore file is a valid path
-    if not Path(args.warning_ignore_file_path).is_file():
+    # Check that a warning ignore file was specified and if so is a valid path
+    if not args.warning_ignore_file_path:
         print(
-            "Warning ignore file does not exist: "
-            f"{args.warning_ignore_file_path}"
+            "Warning ignore file not specified. Continuing without it (no warnings ignored)."
         )
-        return 1
+        files_with_expected_warnings = set()
+    else:
+        if not Path(args.warning_ignore_file_path).is_file():
+            print(
+                f"Warning ignore file does not exist: {args.warning_ignore_file_path}"
+            )
+            return 1
+        with Path(args.warning_ignore_file_path).open(
+            encoding="UTF-8"
+        ) as clean_files:
+            files_with_expected_warnings = {
+                file.strip()
+                for file in clean_files
+                if file.strip() and not file.startswith("#")
+            }
 
     with Path(args.compiler_output_file_path).open(encoding="UTF-8") as f:
         compiler_output_file_contents = f.read()
 
-    with Path(args.warning_ignore_file_path).open(
-        encoding="UTF-8"
-    ) as clean_files:
-        files_with_expected_warnings = {
-            file.strip()
-            for file in clean_files
-            if file.strip() and not file.startswith("#")
-        }
-
-    warnings = extract_warnings_from_compiler_output(
-        compiler_output_file_contents
-    )
+    if args.compiler_output_type == "json":
+        warnings = extract_warnings_from_compiler_output_json(
+            compiler_output_file_contents
+        )
+    elif args.compiler_output_type == "clang":
+        warnings = extract_warnings_from_compiler_output_clang(
+            compiler_output_file_contents
+        )
+
     files_with_warnings = get_warnings_by_file(warnings)
 
     status = get_unexpected_warnings(
-        warnings, files_with_expected_warnings, files_with_warnings
+        files_with_expected_warnings, files_with_warnings
```

Comment:
My IDE says "Expected type 'set[str]', got 'dict[str, list[dict]]' instead" (and for line 232 as well).

Do the functions' hints need updating to `dict`?

Reply:
Good catch. I changed both `get_unexpected_warnings()` and `get_unexpected_improvements()` function hints to the dict